builtin/logical/pki/cert_util.go

```diff
@@ -1513,3 +1515,1039 @@ func stringToOid(in string) (asn1.ObjectIdentifier, error) {
+		// processed before processing this closure.
+
+		var cliques [][]issuerId
+		var cycles [][]issuerId
+		var closure []issuerId
+
+		var cliquesToProcess []issuerId
+		cliquesToProcess = append(cliquesToProcess, issuer)
+
+		for len(cliquesToProcess) > 0 {
+			var node issuerId
+			node, cliquesToProcess = cliquesToProcess[0], cliquesToProcess[1:]
+
+			// Skip potential clique nodes which have already been processed
+			// (either by the topo-sort or by this code).
+			if processed, ok := processedIssuers[node]; ok && processed {
+				continue
+			}
+			var skipNode bool = false
+			for _, closureNode := range closure {
+				if closureNode == node {
+					skipNode = true
+					break
+				}
+			}
+			if skipNode {
+				continue
+			}
+
+			// See if this is a node on a clique and find that clique.
+			cliqueNodes, err := isOnReissuedClique(issuerIdCertMap, subjectIssuerIdsMap, node)
+			if err != nil {
+				// Clique is too large.
+				return toVisit, err
+			}
+
+			// Skip nodes which aren't a clique.
+			if len(cliqueNodes) <= 1 {
+				continue
+			}
+
+			// Add our discovered clique. Note that we avoid duplicate cliques by
+			// the skip logic above. Additionally, we know that cliqueNodes must
+			// be unique and not duplicated with any existing nodes so we can add
+			// all nodes to closure.
+			cliques = append(cliques, cliqueNodes)
+			closure = append(closure, cliqueNodes...)
+
+			// Try and expand the clique to see if there's common cycles around
+			// it.
+			foundCycles, err := findCyclesNearClique(processedIssuers, issuerIdChildrenMap, cliqueNodes)
+			if err != nil {
+				// Cycle is too large.
+				return toVisit, err
+			}
+
+			// We've already handled the clique, so we can continue early if we haven't
+			// found any cycles.
+			if len(foundCycles) == 0 {
+				continue
+			}
+
+			// Assumption: each cycle in foundCycles is in canonical order (see note
+			// below about canonical ordering). Deduplicate these against already
+			// existing cycles and add them to the closure nodes.
+			for _, cycle := range foundCycles {
+				cycles = appendCycleIfNotExisting(cycles, cycle)
+
+				// Now, for each cycle node, we need to find all adjacent cliques.
+				// We do this by finding each child of the cycle and adding it to
+				// the queue. If these nodes aren't cycles, we'll skip them fairly
+				// quickly.
+				for _, node := range cycle {
+					children, ok := issuerIdChildrenMap[node]
+					if !ok {
+						continue
+					}
+
+					for _, child := range children {
+						cliquesToProcess = append(cliquesToProcess, child)
+					}
+
+					// Finally, check if this node is in the closure; if not,
+					// add it.
+					inClosure := false
+					for _, entry := range closure {
+						if entry == node {
+							inClosure = true
+						}
+					}
+					if !inClosure {
+						closure = append(closure, node)
+					}
+				}
+			}
+		}
+
+		// If we lack a closure, go to the next node.
+		if len(closure) == 0 {
+			continue
+		}
```

Comment:
@kitography / @monoidmusician - Does this reasoning sound good?

I do not need a check here to ensure the detected closure has no external, unprocessed dependencies because the starvation condition that we use to call this method (see the `if len(toVisit) == 0` guards prior to this call) guarantee that everything that _could've_ been processed, _should've_ been. 


In particular, consider an external parent that should be processed first, _before_ this closure. This certificate would have to have an (unrelated) issuer that we may not know about, but the correct subject and public key to sign something in the cycle. Without the starvation condition, if we were just calling this randomly, we'd have to ensure that the external parent(s) of this closure was processed before the closure itself. However, because we're starved of nodes to visit (and afterwards, we'll have way too many nodes :-), we know that parent must've already been processed, and that child was unreachable because it had unprocessed parents (due to cycles, i.e., exactly what this code was attempting to prevent against).

Reply:
Per OOB discussion with @MonoidMusician, there's a counter-example here we need to take into account:

Suppose there's two cycles, with an intermediate joining them (i.e., there exists a cert with an issuer in the first cycle, which can validate a node in the second cycle).

This doesn't turn it into a larger cycle, and if the secondary cycle is processed first, its chain will be wrong (because it needs to include the former cycle). So we _do_ need to account for processing state and we _do_ need to take into account other (non-closure) parents.

Reply:
OK, this has been updated.